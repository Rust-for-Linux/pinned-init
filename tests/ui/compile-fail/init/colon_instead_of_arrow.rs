use pin_init::*;

#[pin_data]
struct Foo {
    bar: Bar,
}

#[pin_data]
struct Bar {
    a: usize,
}

impl Bar {
    fn new() -> impl PinInit<Self> {
        pin_init!(Self { a: 42 })
    }
}

impl Foo {
    fn new() -> impl PinInit<Self> {
        pin_init!(Self { bar: Bar::new() })
    }
}

fn main() {}
