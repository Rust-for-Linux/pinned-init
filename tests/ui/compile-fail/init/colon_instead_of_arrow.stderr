error[E0308]: mismatched types
  --> tests/ui/compile-fail/init/colon_instead_of_arrow.rs:21:9
   |
14 |     fn new() -> impl PinInit<Self> {
   |                 ------------------ the found opaque type
...
21 |         pin_init!(Self { bar: Bar::new() })
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         expected `Bar`, found opaque type
   |         arguments to this function are incorrect
   |
   = note:   expected struct `Bar`
           found opaque type `impl pin_init::PinInit<Bar>`
note: function defined here
  --> $RUST/core/src/ptr/mod.rs
   |
   | pub const unsafe fn write<T>(dst: *mut T, src: T) {
   |                     ^^^^^
   = note: this error originates in the macro `$crate::__init_internal` which comes from the expansion of the macro `pin_init` (in Nightly builds, run with -Z macro-backtrace for more info)
